packages/modal/src/ux-modal-service.ts

```diff
@@ -0,0 +1,493 @@
+      for (let layer of this.modalLayers) {
+        if (layer.modal.lock || !layer.modal.outsideDismiss) {
+          concernedLayers = [];
+          continue; // we ignore locks
+        }
+        concernedLayers.push(layer);
+      }
+      // now we have all the layers above the last locked in the concernedLayers array
+      // let's check if the click is outside of any
+      layerloop: for (let layer of concernedLayers) {
+        const modalContentElement = layer.modal.contentElement;
+        for (let element of (event as any).composedPath() ) {
+          if (element === modalContentElement) {
+            continue layerloop; // click is on the modal, do not hide
+          }
+        }
+        layer.modal.dismiss();
+      }
+      return true; // this allow normal behvior with this click for other purposes
+    };
+  }
+
+  public addLayer(modal: UxModal, bindingContext: UxModalBindingContext) {
+    const layerCount = this.modalLayers.push({
+      bindingContext: bindingContext,
+      modal: modal
+    });
+    if (layerCount === 1) {
+      this.setListener();
+    }
+  }
+
+  public getLayer(modal: UxModal): UxModalLayer | undefined {
+    const index = this.modalLayers.map(i => i.modal).indexOf(modal);
+    if (index !== -1) {
+      return this.modalLayers[index];
+    }
+    return undefined;
+  }
+
+  public removeLayer(modal: UxModal) {
+    const index = this.modalLayers.map(i => i.modal).indexOf(modal);
+    if (index !== -1) {
+      this.modalLayers.splice(index, 1);
+    }
+    if (this.modalLayers.length === 0) {
+      this.removeListener();
+    }
+  }
+
+  private setListener() {
+    window.addEventListener('keyup', this);
+    document.addEventListener('click', this.handleDocumentClick);
+  }
+
+  private removeListener() {
+    window.removeEventListener('keyup', this);
+    document.removeEventListener('click', this.handleDocumentClick);
+  }
+
+  public handleEvent(event: KeyboardEvent) {
+    const key = this.getActionKey(event);
+    if (key === undefined) {
+      return;
+    }
+    const activeLayer = this.getLastModal();
+    const keyboard = activeLayer.keyboard;
+    if (key === 'Escape'
+      && (keyboard === true || keyboard === key || (Array.isArray(keyboard) && keyboard.indexOf(key) > -1))) {
+      activeLayer.dismiss();
+    } else if (key === 'Enter' && (keyboard === key || (Array.isArray(keyboard) && keyboard.indexOf(key) > -1))) {
+      activeLayer.ok();
+    }
+  }
+
+  private getActionKey(event: KeyboardEvent): UxModalKeybord | undefined {
+    if ((event.code || event.key) === 'Escape' || event.keyCode === 27) {
+      return 'Escape';
+    }
+    if ((event.code || event.key) === 'Enter' || event.keyCode === 13) {
+      return 'Enter';
+    }
+    return undefined;
+  }
+
+  public getLastLayer(): UxModalLayer {
+    return this.modalLayers[this.modalLayers.length - 1];
+  }
+
+  public getLastModal(): UxModal {
+    return this.getLastLayer().modal;
```

Comment:
If this method is not guaranteed to be called with at least 1 layer, then this line could break

Reply:
good catch. fixed